pkg/controller/controller_binding.go

```diff
@@ -645,119 +652,234 @@ func (c *controller) reconcileServiceBinding(binding *v1beta1.ServiceBinding) er
+		s := "Orphan mitigation successful"
+		setServiceBindingCondition(toUpdate,
+			v1beta1.ServiceBindingConditionReady,
+			v1beta1.ConditionFalse,
+			successOrphanMitigationReason,
+			s)
+	} else {
+		s := "The binding was deleted successfully"
+		setServiceBindingCondition(
+			toUpdate,
+			v1beta1.ServiceBindingConditionReady,
+			v1beta1.ConditionFalse,
+			successUnboundReason,
+			s,
+		)
+		// Clear the finalizer
+		finalizers.Delete(v1beta1.FinalizerServiceCatalog)
+		toUpdate.Finalizers = finalizers.List()
+	}
+
+	toUpdate.Status.ExternalProperties = nil
+	clearServiceBindingCurrentOperation(toUpdate)
+	if _, err := c.updateServiceBindingStatus(toUpdate); err != nil {
+		return err
+	}
+
+	c.recorder.Event(binding, corev1.EventTypeNormal, successUnboundReason, "This binding was deleted successfully")
+	glog.V(5).Info(pcb.Message("Successfully deleted ServiceBinding"))
+
+	return nil
+}
+
+// serviceBindingRequestUnbinding validates and makes the binding request to the broker.
+// Returns if reconciliation should continue and any error produced.
+func (c *controller) serviceBindingRequestUnbinding(binding *v1beta1.ServiceBinding, toUpdate *v1beta1.ServiceBinding, pcb *pretty.ContextBuilder) (bool, error) {
+	glog.V(4).Info(pcb.Message("Going to make request to unbind"))
+	instance, err := c.instanceLister.ServiceInstances(binding.Namespace).Get(binding.Spec.ServiceInstanceRef.Name)
+	if err != nil {
+		s := fmt.Sprintf(
+			`References a non-existent %s "%s/%s"`,
+			pretty.ServiceInstance, binding.Namespace, binding.Spec.ServiceInstanceRef.Name,
+		)
+		glog.Warningf(pcb.Messagef("%s (%s)", s, err))
+		c.recorder.Event(binding, corev1.EventTypeWarning, errorNonexistentServiceInstanceReason, s)
+		setServiceBindingCondition(
+			toUpdate,
+			v1beta1.ServiceBindingConditionReady,
+			v1beta1.ConditionFalse,
+			errorNonexistentServiceInstanceReason,
+			"The binding references an ServiceInstance that does not exist. "+s,
+		)
+		if _, err := c.updateServiceBindingStatus(toUpdate); err != nil {
+			return false, err
+
+		}
+		return false, err
+	}
+
+	if instance.Status.AsyncOpInProgress {
+		s := fmt.Sprintf(
+			`trying to unbind to %s "%s/%s" that has ongoing asynchronous operation`,
+			pretty.ServiceInstance, binding.Namespace, binding.Spec.ServiceInstanceRef.Name,
+		)
+		glog.Info(pcb.Message(s))
+		c.recorder.Event(binding, corev1.EventTypeWarning, errorWithOngoingAsyncOperation, s)
+		setServiceBindingCondition(
+			toUpdate,
+			v1beta1.ServiceBindingConditionReady,
+			v1beta1.ConditionFalse,
+			errorWithOngoingAsyncOperation,
+			errorWithOngoingAsyncOperationMessage,
+		)
+		if _, err := c.updateServiceBindingStatus(toUpdate); err != nil {
+			return false, err
 		}
+		return false, fmt.Errorf("Ongoing Asynchronous operation")
+	}
+
+	if instance.Spec.ClusterServiceClassRef == nil || instance.Spec.ClusterServicePlanRef == nil {
+		// Do not retry if the user wants to cancel/abort a binding request.
```

Comment:
Looks like this out-of-date comment made its way back in.

I think that ultimately here we need to use logic similar to what is done to determine the plan ID for deprovisioning. We need to allow a ServiceBinding to be deleted, with an unbind request sent to the broker, even if the ServiceInstance has been changed to a non-existent plan. That is outside the scope of this PR, however.